### PR TITLE
zone: allow `ListZonesContext` to automatically handle pagination

### PR DESCRIPTION
**Pull request by jacobbednarz**
Updates the `ListZonesContext` method to automatically handle the
pagination provided by the API response to return all zones regardless
of the number of entries.

This will fix cloudflare/terraform-provider-cloudflare#790 which uses
this functionality for filtering the data resource.

Fixes cloudflare/terraform-provider-cloudflare#790

### CODE REVIEW

**Comment by jacobbednarz** [comment]
Confirmed the functionality is working however the tests aren't happy with it. I'll dig into the tests and find out why.

**Comment by jacobbednarz** [comment]
I've updated the stubbed response in 259a650 which was the cause of the failing test 🤦‍♂️

File: zone.go
Comment (patryk) [inline zone.go:457]: It doesn't seem we handle errors returned from channel in any way. We should return them at the end of the function, even a single one is enough. We might also want to cancel pending requests if any goroutine fails this way.

**Review by patryk — CHANGES_REQUESTED** [review]

File: zone.go
Comment (jacobbednarz) [reply zone.go:457]: Is [this](https://github.com/cloudflare/cloudflare-go/pull/534/files#diff-6d1d02aea42947cbb0ae78b9690828a2R477-R478) what you mean? Or did you have something else in mind for the error being returned? I'm not explicitly cancelling the routine but can drop it in if the channel error reporting isn't enough here.

Comment (jacobbednarz) [reply zone.go:457]: Circling back on this one @patryk as we've still got an issue open in the Terraform provider depending on this one.

**Comment by jacobbednarz** [comment]
As this is holding up a Terraform fix (and has sat for a while), let's merge this one in and when we want to fix the error handling you're mentioning we can raise a separate PR to address.

**Pull request merged**
